src/client/dfs/dfs.c

```diff
@@ -1980,17 +1984,101 @@ dfs_umount(dfs_t *dfs)
 		return EINVAL;
 	}
 
+	D_MUTEX_LOCK(&dfs->lock);
+	if (dfs->poh_refcount != 0) {
+		D_ERROR("Pool open handle refcount not 0\n");
+		D_MUTEX_UNLOCK(&dfs->lock);
+		return EBUSY;
+	}
+	if (dfs->coh_refcount != 0) {
+		D_ERROR("Cont open handle refcount not 0\n");
+		D_MUTEX_UNLOCK(&dfs->lock);
+		return EBUSY;
+	}
+	D_MUTEX_UNLOCK(&dfs->lock);
+
 	daos_obj_close(dfs->root.oh, NULL);
 	daos_obj_close(dfs->super_oh, NULL);
 
 	D_FREE(dfs->prefix);
-
 	D_MUTEX_DESTROY(&dfs->lock);
 	D_FREE(dfs);
 
 	return 0;
 }
 
+int
+dfs_pool_get(dfs_t *dfs, daos_handle_t *poh)
+{
+	if (dfs == NULL || !dfs->mounted)
+		return EINVAL;
+
+	D_MUTEX_LOCK(&dfs->lock);
+	dfs->poh_refcount++;
+	D_MUTEX_UNLOCK(&dfs->lock);
```

Comment:
Could we avoid a lock here by using atomics? Although this probably won't be called many times

Reply:
how would you handle the case in put where you want to check the value before taking the refcount, atomically?

```
	if (dfs->coh_refcount <= 0) {
		D_ERROR("Invalid cont handle refcount\n");
		D_MUTEX_UNLOCK(&dfs->lock);
		return EINVAL;
	}
	dfs->coh_refcount--;
```

Reply:
Ah, right. It would require at least two operations. Locking is simpler here

Reply:
For infrequent operations then locks are much simpler.